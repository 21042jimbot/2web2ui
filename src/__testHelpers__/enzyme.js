export const findButton = (wrapper, buttonLabel) =>
  wrapper.findWhere(
    (node) => node.type() === 'button' && new RegExp(buttonLabel).test(node.text())
  );
