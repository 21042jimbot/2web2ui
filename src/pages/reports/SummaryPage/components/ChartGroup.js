import React, { Component } from 'react';
import _ from 'lodash';
import { getDayLines, getLineChartFormatters } from 'helpers/chart';
import LineChart from './LineChart';
import METRICS_UNIT_CONFIG from 'config/metrics-units';
const DEFAULT_UNIT = 'number';

function getUniqueUnits(metrics) {
  return _.uniq(metrics.map(({ unit = DEFAULT_UNIT }) => unit));
}

export default class ChartGroup extends Component {

  createDayReferenceLines() {
    const { data, precision } = this.props;

    return getDayLines(data, precision).map(({ ts }) => ({
      key: ts,
      x: ts,
      stroke: '#bbb',
      strokeWidth: 2
    }));
  }

  render() {
    const { data = [], metrics, loading, precision } = this.props;

    if (!data.length || !metrics) {
      return null;
    }

    const formatters = getLineChartFormatters(precision);
    const referenceLines = this.createDayReferenceLines();
    const charts = getUniqueUnits(metrics).map((unit) => ({
      metrics: metrics.filter((metric) => metric.unit === unit),
      ...METRICS_UNIT_CONFIG[unit]
    }));

    return (
      <div>
        {charts.map((chart, i) => <LineChart
          key={`chart=${i}`}
          syncId='summaryChart'
          data={data}
          lines={chart.metrics.map(({ name, label, stroke }) => ({
            key: name,
            dataKey: name,
            name: label,
            stroke: loading ? '#f8f8f8' : stroke
          }))}
          {...formatters}
          yTickFormatter={chart.yAxisFormatter}
          tooltipValueFormatter={chart.yAxisFormatter}
          referenceLines={referenceLines}
          showXAxis={i === charts.length - 1}
        />)}
      </div>
    );
  }

}
