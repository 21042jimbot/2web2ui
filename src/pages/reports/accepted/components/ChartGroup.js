import React, { Component } from 'react';
import { connect } from 'react-redux';
import _ from 'lodash';
import { Grid, Panel } from '@sparkpost/matchbox';
import { Loading, PieChart } from 'src/components';
import { generateColors } from 'src/helpers/color';
import styles from './ChartGroup.module.scss';

// Chart color palette generated from:
const primaryColor = '#0097b3';

export class ChartGroup extends Component {
  state = {
    hoveredItem: null,
    active: null
  };

  /**
   * Handles mouse over event for LegendItems and BounceChart
   * @param  {Object} e - Recharts synthetic event - behavior mimiced from Legend
   * @param  {string} hoverSet - 'primary' | 'secondary'
   */
  handleMouseOver = (e, hoverSet) => {
    const { deliveries } = this.props;
    const { active } = this.state;
    const { name, count } = e;

    let dataSet = deliveries;

    if (active) {
      dataSet = active.children;
    }

    const hoveredItem = {
      name,
      count,
      index: _.findIndex(dataSet, { name }),
      dataSet: hoverSet
    };

    this.setState({ hoveredItem });
  }

  handleMouseOut = () => {
    this.setState({ hoveredItem: null });
  }

  handleClick = ({ name, children, count }) => {
    if (children) {
      this.setState({
        active: { name, children, count },
        hoveredItem: null
      });
    }
  }

  handleBreadcrumb = () => {
    this.setState({ active: null });
  }

  getLabelProps = () => {
    const { aggregates } = this.props;
    const { hoveredItem } = this.state;

    const getRate = (n) => `${(100 * n).toFixed(2)}%`;

    return hoveredItem
      ? { name: hoveredItem.name, value: getRate(hoveredItem.count / aggregates.count_accepted) }
      : { name: 'Accepted Rate', value: getRate(aggregates.count_accepted / aggregates.count_targeted) };
  }

  getLegendHeaderData = () => {
    const { aggregates } = this.props;
    const { active } = this.state;

    // Header with breadcrumb & active data
    if (active) {
      return [
        { name: 'All Accepted', breadcrumb: true, onClick: this.handleBreadcrumb },
        { name: 'All Targeted', count: aggregates.count_targeted },
        { name: active.name, count: active.count }
      ];
    }

    // Default header
    return [
      { name: 'Targeted', count: aggregates.count_targeted },
      { name: 'All Accepted', count: aggregates.count_accepted }
    ];
  }

  // Gets primary and secondary data for BounceChart & Legend
  getData = () => {
    const { deliveries } = this.props;
    const { active } = this.state;

    let primaryData = deliveries;

    if (active) {
      primaryData = active.children;
    }

    return {
      primaryData: generateColors(primaryData, primaryColor)
    };
  }

  render() {
    const { loading } = this.props;

    if (loading) {
      return <Panel title='Accepted Attempt Rates' sectioned className={styles.LoadingPanel}><Loading /></Panel>;
    }

    return (
      <Panel title='Accepted Attempt Rates' sectioned>
        <Grid>
          <Grid.Column xs={12} lg={5}>
            <div className={styles.ChartWrapper}>
              <PieChart.Chart
                {...this.getData()}
                hoveredItem={this.state.hoveredItem}
                onMouseOver={this.handleMouseOver}
                onMouseOut={this.handleMouseOut}
                onClick={this.handleClick} />
              <PieChart.ActiveLabel {...this.getLabelProps()}/>
            </div>
          </Grid.Column>
          <Grid.Column xs={12} lg={7}>
            <PieChart.Legend
              headerData={this.getLegendHeaderData()}
              {...this.getData()}
              hoveredItem={this.state.hoveredItem}
              onMouseOver={this.handleMouseOver}
              onMouseOut={this.handleMouseOut}
              onClick={this.handleClick} />
          </Grid.Column>
        </Grid>
      </Panel>
    );
  }
}

const mapStateToProps = ({ acceptedReport }) => ({
  loading: acceptedReport.aggregatesLoading,
  deliveries: acceptedReport.deliveries,
  aggregates: acceptedReport.aggregates
});
export default connect(mapStateToProps, {})(ChartGroup);
