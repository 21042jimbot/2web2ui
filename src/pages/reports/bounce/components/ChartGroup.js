import React, { Component } from 'react';
import { connect } from 'react-redux';
import _ from 'lodash';

import { Grid, Panel } from '@sparkpost/matchbox';
import ActiveLabel from './ActiveLabel';
import BounceChart from './BounceChart';
import Legend from './Legend';
import { Loading } from 'src/components';
import { generateColors } from 'src/helpers/bounce';

import styles from './ChartGroup.module.scss';

// Chart color palette generated from:
const primaryColor = '#DB2F3D';
const secondaryColor = '#37aadc';

export class ChartGroup extends Component {
  state = {
    hoveredItem: null,
    active: null
  };

  handleMouseOver = (e, hoverSet) => {
    const { categories, types } = this.props;
    const { active } = this.state;
    const { name, count } = e;

    let dataSet = hoverSet === 'primary' ? categories : types;

    if (active) {
      dataSet = active.children;
    }

    const hoveredItem = {
      name,
      count,
      index: _.findIndex(dataSet, { name }),
      dataSet: hoverSet
    };

    this.setState({ hoveredItem });
  }

  handleMouseOut = () => {
    this.setState({ hoveredItem: null });
  }

  handleClick = ({ name, children, count }) => {
    if (children) {
      this.setState({
        active: { name, children, count },
        hoveredItem: null
      });
    }
  }

  handleBreadcrumb = () => {
    this.setState({ active: null });
  }

  getLabelProps = () => {
    const { aggregates } = this.props;
    const { hoveredItem } = this.state;

    const getRate = (n) => `${(100 * n).toFixed(2)}%`;

    return hoveredItem
      ? { name: hoveredItem.name, value: getRate(hoveredItem.count / aggregates.countBounce) }
      : { name: 'Bounce Rate', value: getRate(aggregates.countBounce / aggregates.countTargeted) };
  }

  getLegendHeaderData = () => {
    const { aggregates } = this.props;
    const { active } = this.state;

    // Header with breadcrumb & active data
    if (active) {
      return [
        { name: 'All Bounces', breadcrumb: true, onClick: this.handleBreadcrumb },
        { name: active.name, count: active.count }
      ];
    }

    // Default header
    return [
      { name: 'Targeted', count: aggregates.countTargeted },
      { name: 'All Bounces', count: aggregates.countBounce }
    ];
  }

  // Gets primary and secondary data for BounceChart & Legend
  getData = () => {
    const { categories, types } = this.props;
    const { active } = this.state;

    let primaryData = categories;
    let secondaryData = types;

    if (active) {
      primaryData = active.children;
      secondaryData = null;
    }

    return {
      primaryData: generateColors(primaryData, primaryColor),
      secondaryData: secondaryData && generateColors(secondaryData, secondaryColor)
    };
  }

  render() {
    const { loading, aggregates } = this.props;

    if (loading || !aggregates) {
      return <Panel title='Bounce Rates' sectioned className={styles.LoadingPanel}><Loading /></Panel>;
    }

    return (
      <Panel title='Bounce Rates' sectioned>
        <Grid>
          <Grid.Column xs={12} lg={5}>
            <div className={styles.ChartWrapper}>
              <BounceChart
                {...this.getData()}
                hoveredItem={this.state.hoveredItem}
                handleMouseOver={this.handleMouseOver}
                handleMouseOut={this.handleMouseOut}
                handleClick={this.handleClick} />
              <ActiveLabel {...this.getLabelProps()}/>
            </div>
          </Grid.Column>
          <Grid.Column xs={12} lg={7}>
            <Legend
              headerData={this.getLegendHeaderData()}
              {...this.getData()}
              handleMouseOver={this.handleMouseOver}
              handleMouseOut={this.handleMouseOut}
              handleClick={this.handleClick} />
          </Grid.Column>
        </Grid>
      </Panel>
    );
  }
}

const mapStateToProps = ({ bounceReport }) => ({
  loading: bounceReport.aggregatesLoading || bounceReport.categoriesLoading,
  aggregates: bounceReport.aggregates,
  categories: bounceReport.categories,
  types: bounceReport.types
});
export default connect(mapStateToProps, {})(ChartGroup);
