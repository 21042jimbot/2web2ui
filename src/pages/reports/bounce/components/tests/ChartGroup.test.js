import React from 'react';
import { ChartGroup } from '../ChartGroup';
import { shallow } from 'enzyme';

describe('ChartGroup: ', () => {

  const props = {
    categories: [{ name: 'cats', count: 1, children: [{ name: 'child' }]}],
    types: [{ name: 'types', count: 2, children: [{ name: 'child' }]}],
    aggregates: {
      countTargeted: 100,
      countBounce: 50
    }
  };

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<ChartGroup {...props} />);
  });

  it('should render', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render loading', () => {
    wrapper.setProps({ loading: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('should render label when hovering', () => {
    wrapper.instance().handleMouseOver({ name: 'cats', count: 2 });
    wrapper.update();
    expect(wrapper.find('ActiveLabel')).toMatchSnapshot();
  });

  it('should handle mouse out', () => {
    wrapper.instance().handleMouseOver({ name: 'cats', count: 2 });
    expect(wrapper.instance().state).toMatchSnapshot();
    wrapper.instance().handleMouseOut();
    expect(wrapper.instance().state).toMatchSnapshot();
  });

  it('should render correctly after click', () => {
    wrapper.instance().handleClick({ name: 'cats', count: 3, children: [{ name: 'child' }]});
    wrapper.update();
    expect(wrapper.find('PieChart.Legend')).toMatchSnapshot();
    expect(wrapper.find('PieChart.Chart')).toMatchSnapshot();
  });

  it('should handle breadcrumb', () => {
    wrapper.instance().handleClick({ name: 'cats', count: 3, children: [{ name: 'child' }]});
    expect(wrapper.instance().state).toMatchSnapshot();
    wrapper.instance().handleBreadcrumb();
    expect(wrapper.instance().state).toMatchSnapshot();
  });
});
