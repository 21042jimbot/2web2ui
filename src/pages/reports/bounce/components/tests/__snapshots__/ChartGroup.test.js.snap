// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartGroup:  should handle breadcrumb 1`] = `
Object {
  "active": Object {
    "children": Array [
      Object {
        "name": "child",
      },
    ],
    "count": 3,
    "name": "cats",
  },
  "hoveredItem": null,
}
`;

exports[`ChartGroup:  should handle breadcrumb 2`] = `
Object {
  "active": null,
  "hoveredItem": null,
}
`;

exports[`ChartGroup:  should handle mouse out 1`] = `
Object {
  "active": null,
  "hoveredItem": Object {
    "count": 2,
    "dataSet": undefined,
    "index": -1,
    "name": "cats",
  },
}
`;

exports[`ChartGroup:  should handle mouse out 2`] = `
Object {
  "active": null,
  "hoveredItem": null,
}
`;

exports[`ChartGroup:  should render 1`] = `
<Panel
  sectioned={true}
  title="Bounce Rates"
>
  <Grid>
    <Grid.Column
      lg={5}
      xs={12}
    >
      <div>
        <BounceChart
          handleClick={[Function]}
          handleMouseOut={[Function]}
          handleMouseOver={[Function]}
          hoveredItem={null}
          primaryData={
            Array [
              Object {
                "children": Array [
                  Object {
                    "name": "child",
                  },
                ],
                "count": 1,
                "fill": "hsl(355.1, 70.5%, 52.2%)",
                "name": "cats",
              },
            ]
          }
          secondaryData={
            Array [
              Object {
                "children": Array [
                  Object {
                    "name": "child",
                  },
                ],
                "count": 2,
                "fill": "hsl(198.20000000000005, 70.2%, 53.9%)",
                "name": "types",
              },
            ]
          }
        />
        <ActiveLabel
          name="Bounce Rate"
          value="50.00%"
        />
      </div>
    </Grid.Column>
    <Grid.Column
      lg={7}
      xs={12}
    >
      <Legend
        handleClick={[Function]}
        handleMouseOut={[Function]}
        handleMouseOver={[Function]}
        headerData={
          Array [
            Object {
              "count": 100,
              "name": "Targeted",
            },
            Object {
              "count": 50,
              "name": "All Bounces",
            },
          ]
        }
        hoveredItem={null}
        primaryData={
          Array [
            Object {
              "children": Array [
                Object {
                  "name": "child",
                },
              ],
              "count": 1,
              "fill": "hsl(355.1, 70.5%, 52.2%)",
              "name": "cats",
            },
          ]
        }
        secondaryData={
          Array [
            Object {
              "children": Array [
                Object {
                  "name": "child",
                },
              ],
              "count": 2,
              "fill": "hsl(198.20000000000005, 70.2%, 53.9%)",
              "name": "types",
            },
          ]
        }
      />
    </Grid.Column>
  </Grid>
</Panel>
`;

exports[`ChartGroup:  should render correctly after click 1`] = `
<Legend
  handleClick={[Function]}
  handleMouseOut={[Function]}
  handleMouseOver={[Function]}
  headerData={
    Array [
      Object {
        "breadcrumb": true,
        "name": "All Bounces",
        "onClick": [Function],
      },
      Object {
        "count": 3,
        "name": "cats",
      },
    ]
  }
  hoveredItem={null}
  primaryData={
    Array [
      Object {
        "fill": "hsl(355.1, 70.5%, 52.2%)",
        "name": "child",
      },
    ]
  }
  secondaryData={null}
/>
`;

exports[`ChartGroup:  should render correctly after click 2`] = `
<BounceChart
  handleClick={[Function]}
  handleMouseOut={[Function]}
  handleMouseOver={[Function]}
  hoveredItem={null}
  primaryData={
    Array [
      Object {
        "fill": "hsl(355.1, 70.5%, 52.2%)",
        "name": "child",
      },
    ]
  }
  secondaryData={null}
/>
`;

exports[`ChartGroup:  should render label when hovering 1`] = `
<ActiveLabel
  name="cats"
  value="4.00%"
/>
`;

exports[`ChartGroup:  should render loading 1`] = `
<Panel
  sectioned={true}
  title="Bounce Rates"
>
  <Component />
</Panel>
`;
